File: web/src/entities/spike.ts
import { ImmutableVec2 } from '../math.js';
import { Entity } from '../entity.js';
import { Resource } from '../resource.js';
import type { GameImage } from '../resource.js';

export class Spike extends Entity {
  private readonly spikeTile: GameImage;

  constructor() {
    super();
    this.setSize(new ImmutableVec2(10, 10));
    this.spikeTile = Resource.getBitmap('data/images/tileset1.bmp').subImage(70, 0, 10, 10);
  }

  getLayer(): number { return 3; }

  override draw(ctx: CanvasRenderingContext2D, offsetX: number, offsetY: number, _layer: number): void {
    this.spikeTile.draw(ctx,
      Math.round(offsetX + this.getPosition().x - this.getHalfSize().x),
      Math.round(offsetY + this.getPosition().y - this.getHalfSize().y));
  }

  override update(): void {
    if (this.myRoom.getHero().Collides(this)) {
      this.myRoom.getHero().kill();
    }
  }
}

File: wii/src/entities/spike.cpp
#include "entities/spike.h"
#include "entities/hero.h"
#include "resource.h"
#include "room.h"
#include <cmath>

Spike::Spike() {
    setSize(Vec2(10, 10));
    spikeTile_ = Resource::getBitmap("data/images/tileset1.bmp").subImage(70, 0, 10, 10);
}

void Spike::draw(SDL_Renderer* renderer, int offsetX, int offsetY, int /*layer*/) {
    int x = static_cast<int>(std::round(offsetX + position_.x - halfSize_.x));
    int y = static_cast<int>(std::round(offsetY + position_.y - halfSize_.y));
    spikeTile_.draw(renderer, x, y);
}

void Spike::update() {
    if (room_->getHero()->Collides(*this)) {
        room_->getHero()->kill();
    }
}
